### PR TITLE
PAYARA-3438: Improve usability MP Health endpoint as readinessProbe

### PR DESCRIPTION
**Pull request by rdebusscher**
… for k8s

### CODE REVIEW

**Comment by svendiedrichsen** [comment]
@rdebusscher Looks as if this only works if you have a single app to deploy or how do you specify the target number of deployed applications?

**Comment by rdebusscher** [comment]
@svendiedrichsen Yes it works optimally for a single app in Payara Micro. Having multiple apps on 1 Payara Micro doesn't make sense if you like to optimize the scale for each app differently.

**Comment by svendiedrichsen** [comment]
@rdebusscher But the MP Health endpoint ist available in Payara full as well. And you can run this with multiple apps in docker/k8s. This can't be used in such an environment for a readiness probe then.

How about providing a system property at server startup with the expected number of apps? This could if absent default to 1. Otherwise why bother with a list of app names? This could be a boolean.

File: appserver/payara-appserver-modules/microprofile/healthcheck/src/main/java/fish/payara/microprofile/healthcheck/HealthCheckService.java [hunk 1]
```diff
@@ -191,6 +202,12 @@ public void performHealthChecks(HttpServletResponse response) throws IOException
             }
         }
 
+        // No applications (yet), server is not ready.
+        if (applicationsLoaded.isEmpty()) {
+            // Application is not yet deployed
+            healthCheckResponses.add(HealthCheckResponse.builder().down().build());
```
Comment (Pandrex247) [inline appserver/payara-appserver-modules/microprofile/healthcheck/src/main/java/fish/payara/microprofile/healthcheck/HealthCheckService.java]: Is this valid? It should at least have a name.

File: appserver/payara-appserver-modules/microprofile/healthcheck/src/main/java/fish/payara/microprofile/healthcheck/HealthCheckService.java [hunk 2]
```diff
@@ -112,6 +114,15 @@ public void event(Event event) {
             if (appInfo != null) {
                 healthChecks.remove(appInfo.getName());
                 applicationClassLoaders.remove(appInfo.getName());
+                applicationsLoaded.remove(appInfo.getName());
+            }
+        }
+
+        // Keep track of all deployed applications.
+        if (event.is(Deployment.APPLICATION_STARTED)) {
+            ApplicationInfo appInfo = Deployment.APPLICATION_STARTED.getHook(event);
```
[the remark this answers is not in the document]
Comment (rdebusscher) [reply appserver/payara-appserver-modules/microprofile/healthcheck/src/main/java/fish/payara/microprofile/healthcheck/HealthCheckService.java]: Application_loaded is too early, the endpoints are not available at that time.
Application must be fully operational

**Comment by smillidge** [comment]
What do we need to do to get this in?

File: appserver/payara-appserver-modules/microprofile/healthcheck/src/main/java/fish/payara/microprofile/healthcheck/HealthCheckService.java (hunk 1 above)
Comment (rdebusscher) [reply appserver/payara-appserver-modules/microprofile/healthcheck/src/main/java/fish/payara/microprofile/healthcheck/HealthCheckService.java]: I can't find anything in the spec that the name is required. But I added a name.

**Comment by arjantijms** [comment]
Jenkins test

**Comment by arjantijms** [comment]
Jenkins test

**Pull request merged**